Update Effect System ADR (#679)
informalsystems/quint: doc/adr004-effect-system.md

which are independent of each other. Here are some examples of inferred effects
for some definitions:

```
var x: int
var y: int

pure def P(a) = a + 1 // (Read[r1] & Temporal[t1]) => Read[r1] & Temporal[t1]
def S(a) = y + a // (Read[r1] & Temporal[t1]) => Read['y', r1] & Temporal[t1]

action A1 = x' = x + 1 // Read['x'] & Update['x']
action A2 = x' = S(1) // Read['y'] & Update['x']
```

In these examples, keywords `pure def`, `def` and `action` are providing effect
information to be checked against the inferred effect. This information is not
used for inference.

### How restrictive should `pure` definitions be?

We have two alternatives for the `pure` keyword:

1. Only allow Pure parameters and Pure result
2. Allow parameters with Read effects as long as the resulting effect doesn't
   read any other variables

Option 1 may be too restrictive and barely used, since, for most operators, it's
desirable that they can be applied to state variables (i.e. `+` in `x + 1` if
`x` is a state variable).

### Decision

We have decided to implement option 2 for all modes. According to this option, a
definition is considered pure if it does not add any read effect on top of the

> effects that the parameters may carry. A definition is considered `def` if it

Reviewer: ```suggestion
effects that the parameters may carry. A definition is considered `pure` if it
```

Author: This is actually meant to be `def`. Since the option 2 was implemented for all modes, not only pure, I gave a description of the `def` mode as well. I'll update this to clarify that this is another example, as I agree it is a bit confusing in the current structure.

introduces a new `Read` effect, but does not introduce any `Update` or
`Temporal` effects. The same principle applies to other types of definitions.

### Notation

```
Identifiers v, c, op

> Effects E, Ei ::= e | Read[entity] | Update[entity] | Temporal[entity] | Pure | E & E | (E0, ..., EN) => E

Author: `e` is suppose to be an effect variable. I will change the symbol for `Expression` to `expr`

> Entities V ::= v | 'x' | v1, ..., vn

Author: `v1, ..., vn` was suppose to be `V1, ..., Vn`, so `V` is an entity that can be replaced by entity variables `v` or state variables `'x'`

```

A state variable is just a name that is exactly the name of a state variable in
the spec. It also carries an optional reference to the expression on which the
state variable occurred, which is used for error reporting.

```typescript
/* A state variable */
export interface StateVariable {
  /* The variable name */
  name: string,
  /* The id of the expression on which the state variable ocurred */
  reference: bigint
}
```

### Effects

An effect can also have three different forms: concrete, arrow and variable.

```typescript
/*
 * The effect of a Quint expression, regarding which state entities are read
 * and/or updated
 */
export type Effect =
  | ConcreteEffect
  | ArrowEffect
  | EffectVariable
```

The most basic representation of effects is a concrete effect, and it is
structured as a list of effect components. An effect component has a kind and an
entity. The kind can be `read`, `update` or `temporal`, and the entity is the
entity the effect acts upon.

```typescript
/* Kinds for concrete effect components */
export type ComponentKind = 'read' | 'update' | 'temporal'

/* A concrete effect component, specifying a kind and the entity it acts upon */

> export interface EffectComponent { kind: ComponentKind, entity: Entity }

Reviewer 2: I'm not sure why we need the `EffectComponent` now. Iiuc, it's just a product of an effect kind and entity, so maybe using a tuple would suffice and reduce the number of abstract nouns we have to track by one? Just thinking out loud... :)

Author: We have to define a name for the field in `ConcreteEffect` that carries this, which is now `components`. So I don't think we can reduce the number, and this might help clarify it.

/* Concrete atomic effects specifying which kinds of effects act upon which entities */
export interface ConcreteEffect { kind: 'concrete', components: EffectComponent[] }

```

The effect `Read['x', 'y'] & Update[v]` would be represented as:

```typescript
{
  kind: 'concrete',
  components: [
    { kind: 'read', entity: { kind: 'concrete', stateVariables: [{ name: 'x', reference: 1n }, { name: 'y', reference: 2n }] } },
    { kind: 'update', entity: { kind: 'variable', name: 'v' } },
  ],
}
```

Effect signatures for operators may depend on its parameters. For that purpose,
we also define arrow effects, which carry a list of effects for the respective
parameters of an operator and the resulting effect.

```typescript
/* Arrow effects for expressions with effects depending on parameters */
export interface ArrowEffect { kind: 'arrow', params: Effect[], result: Effect }
```

> Just like with entities, an effect can also be a variable, which can be
> substituted with another effect during inferrence. Those are the most general

Reviewer 2: I think the phasing 'an effect can also be a variable' sets up a point of possoble conceptual and terminological confusion. Rather, there is a class of variables that range over effects (and a different class that range over "entities", and "state variables" aren't actually variables that range over anything, lol). So how about something like:

```suggestion
These is a separate class of variables that range over effects, which can be
substituted with another effect during inference. Those are the most general
```

> kinds of effects, and will be used when no effect information for a given
> expression is available, i.e. `def foo(x) = x` will have a `(e) => e` effect
> signature, where `e` is an effect variable, since there are no constraints on
> the effect of `x`. An effect variable is equivalent to `Read[v1] & Update[v2] &
> Temporal[v3]`, where `v1`, `v2` and `v3` are entity variables.

Reviewer 2: I think that possibly part of the conceptual confusion i am hitting in this paragraph results from not keeping a clear separation between the ts representation forming our meta language, and the object language? So, `EffectVariable` is one of the types comprising the `Effect` union type`, but an effect variable *stands for* but *is not itself* an effect. Does this make sense?

This would suggest a rephrasing like:

```suggestion
When no effect information for a given expression is available, the effect is left
undetermined, as indicated by an effect variable. I.e. `def foo(x) = x` will have be inferred to have the effect signature `(e) => e`, where `e` is an effect variable, since there are no constraints on the effect of `x`. An **effect variable** `e` is equivalent to the effect `Read[v1] & Update[v2] & Temporal[v3]`, where `v1`, `v2` and `v3` are _entity variables_.
```

```typescript

> /* A variable representing some effect */

Reviewer 2: ```suggestion
/* A variable standing for an unknown effect */
```

export interface EffectVariable { kind: 'variable', name: string }
```

### Effect Schemes

Both effects and entities can be universally quantified, and we need to keep

> track of which of them are. Effect schemes are responsible to keep track of that

Reviewer 2: ```suggestion
track of which of them are. Effect schemes are responsible for keeping track of that
```

information.

```typescript
/*

>  * An effect scheme, listing which effect variables (names referring to effects)
>  * and entity variables (names refering to entities) are universally quantified

Reviewer 2: I think this also confuses levels a bit. IIUC, the names here refer to variables, which stand for unknown effects, they do not refer to effects directly.

```suggestion
 * An effect scheme, listing which effect variables (represented by the names of effect variables)
 * and entity variables (represented by the names of entity variables) are universally quantified
```

 * in the effect
 */
export type EffectScheme = {
  /* The effect */
  effect: Effect,

>   /* Universally quantified names refering to Effects */
>   effectVariables: Set<string>,
>   /* Universally quantified names refering to Entities */

Reviewer 2: ```suggestion
  /* The names of universally quantified Effect variables */
  effectVariables: Set<string>,
  /* The names of universally quantified Entity variables */
```

  entityVariables: Set<string>,
}
```

An effect like `(e25, Read[v22]) => Read['x', v22])` with both `e25` and `v22`
being universally quantified would be represented as:

```typescript
{
  effect: ...,
  effectVariables: new Set(['e25']),
  entityVariables: new Set(['v22']),
}
```

And pretty printed as:
```
∀ e0, v0 . (e0, Read[v0]) => Read['x', v0]
```

### Unification

Unification is pretty standard. It recursively looks at effect variables and
entity variables to bind. The most complicated part must be the unification of
unions of variables, which requires unifying sets. Since that is too complicated
to implement, we use some workarounds to avoid having to do that in the few

> cases where we would need to. Those workarounds revolve on reasoning about one

Reviewer 2: ```suggestion
cases where we would need to. Those workarounds rely on reasoning about one
```

Or 'revolve around'

of the unions as a single new variable, which is done by introducing a new
hashed variable, which is then binded to the other union. A hashed variable for

> the union `r1, r2, r3` is `r1#r2#r3`. These hashing procedure is also used to

Reviewer 2: ```suggestion
the union `r1, r2, r3` is `r1#r2#r3`. This hashing procedure is also used to
```

unpack tuples into lambdas that take more than one argument.

### Inference rules

Inferring names: variables have effect `Read[<variable_name>]` (unless they are

> used in as targets of assignment, where their resulting effect will be inferred

Reviewer 2: ```suggestion
used as targets of assignment, where their resulting effect will be inferred
```

/* A concrete effect component, specifying a kind and the entity it acts upon */
export interface EffectComponent { kind: ComponentKind, entity: Entity }

/* Concrete atomic effects specifying which kinds of effects act upon which entities */
export interface ConcreteEffect { kind: 'concrete', components: EffectComponent[] }

```

The effect `Read['x', 'y'] & Update[v]` would be represented as:

```typescript
{
  kind: 'concrete',
  components: [
    { kind: 'read', entity: { kind: 'concrete', stateVariables: [{ name: 'x', reference: 1n }, { name: 'y', reference: 2n }] } },
    { kind: 'update', entity: { kind: 'variable', name: 'v' } },
  ],
}
```

The effect signature of an operator may depend on its parameters. For that reason,
we also define arrow effects, which carry a list of effects for the respective
parameters of an operator and the effect resulting from the operator's application.

```typescript
/* Arrow effects for expressions with effects depending on parameters */
export interface ArrowEffect { kind: 'arrow', params: Effect[], result: Effect }
```

These is a separate class of variables that range over effects, which can be

> substituted with another effect during inference. Those are the most general
> When no effect information for a given expression is available, the effect is left

Reviewer 2: Looks like part of the sentence was lost in applying suggestions:

```suggestion
substituted with another effect during inference. Effect variables are the most general kind of effects.
When no effect information for a given expression is available, the effect is left
```

We have decided to implement option 2 for all modes. According to this option, a
definition is considered pure if it does not add any read effect on top of the
effects that the parameters may carry. A definition is considered (impure) `def` if it
introduces a new `Read` effect, but does not introduce any `Update` or
`Temporal` effects. The same principle applies to other types of definitions.

### Notation

```
StateVariable 'x', 'y', 'z'
ConstantExpression c
Operator op
EntityVariable v
Entities V, Vi ::= v | 'x' | V1, ..., Vn
EffectVariable e
Effect E, Ei ::= e | Read[V] | Update[V] | Temporal[V] | Pure | E & E | (E0, ..., EN) => E
Expression expr, expri ::= any Quint expression
Context Γ ::= lookup tables and previously inferred effects

> Substitution S, Si ::= {e ↦ E} | {v ↦ V} | S ∪ S

Reviewer 2: This totally clears up my confusions. Thanks!
